moxin-backend/src/backend_impls.rs

```diff
@@ -0,0 +1,716 @@
+                chat_completion_message: None,
+            }
+        }
+
+        fn init_request(&mut self) -> Result<(), ()> {
+            if let Ok((req, tx)) = self.request_rx.recv() {
+                // Init current_req
+                if !req.stream.unwrap_or_default() {
+                    self.chat_completion_message = Some(Vec::with_capacity(
+                        (req.max_tokens.unwrap_or(512) * 8) as usize,
+                    ))
+                }
+                *self.current_req.get_mut() = serde_json::to_vec(&req).unwrap();
+                self.current_req.set_position(0);
+                self.token_tx = Some(tx);
+                Ok(())
+            } else {
+                Err(())
+            }
+        }
+
+        pub fn read_data(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+            let n = self.current_req.read(buf)?;
+            if n == 0 {
+                self.current_req.get_mut().clear();
+                self.current_req.set_position(0);
+            }
+            Ok(n)
+        }
+
+        fn send_output(&mut self, output: Result<&[u8], TokenError>) -> bool {
+            if self.token_tx.is_none() {
+                return false;
+            }
+
+            match (output, &mut self.chat_completion_message) {
+                (Ok(token), Some(chat_completion_message)) => {
+                    chat_completion_message.extend_from_slice(token);
+                }
+                (Ok(token), None) => {
+                    let _ =
+                        self.token_tx
+                            .as_mut()
+                            .unwrap()
+                            .send(Ok(ChatResponse::ChatResponseChunk(ChatResponseChunkData {
+                                id: String::new(),
+                                choices: vec![ChunkChoiceData {
+                                    finish_reason: None,
+                                    index: 0,
+                                    delta: MessageData {
+                                        content: String::from_utf8_lossy(token).to_string(),
+                                        role: Role::Assistant,
+                                    },
+                                    logprobs: None,
+                                }],
+                                created: 0,
+                                model: String::new(),
+                                system_fingerprint: String::new(),
+                                object: "chat.completion.chunk".to_string(),
+                            })));
+                }
+                (Err(token_error), chat_completion_message) => {
+                    if let Some(chat_completion_message) = chat_completion_message.take() {
+                        let _ = self.token_tx.as_mut().unwrap().send(Ok(
+                            ChatResponse::ChatFinalResponseData(ChatResponseData {
+                                id: String::new(),
```

Comment:
Should we autogenerate an id for the chat interaction? Maybe it can be done a bit later just to be fully compatible with the OpenAI format

Reply:
I don't understand the generation format of the ID field in OpenAI, can I use a random UUID instead?

Reply:
Using random UUID seems OK to me 👍